upnpdiscovery/service.go

```diff
@@ -0,0 +1,229 @@
+}
+
+// convertqueryConfigToDeviceQuery pulls out the device query from a queryConfig and creates a map of queries to endpoints.
+// The map is needed to add the endpoints to a host when creating the RTSP url.
+func convertqueryConfigToDeviceQuery(queryCfgs []queryConfig) ([]viamupnp.DeviceQuery, map[viamupnp.DeviceQuery][]string) {
+	var queries []viamupnp.DeviceQuery
+	endpointMap := make(map[viamupnp.DeviceQuery][]string)
+
+	for _, queryCfg := range queryCfgs {
+		upnpQuery := viamupnp.DeviceQuery{
+			ModelName:    queryCfg.ModelName,
+			Manufacturer: queryCfg.Manufacturer,
+			SerialNumber: queryCfg.SerialNumber,
+			Network:      queryCfg.Network,
+		}
+		queries = append(queries, upnpQuery)
+		if len(queryCfg.Endpoints) > 0 {
+			endpointMap[upnpQuery] = queryCfg.Endpoints
+		}
+	}
+	return queries, endpointMap
+}
+
+// DiscoverResources discovers different rtsp cameras that use onvif.
+func (dis *upnpDiscovery) DiscoverResources(ctx context.Context, extra map[string]any) ([]resource.Config, error) {
+	cams := []resource.Config{}
+
+	discoverQueries := dis.queries
+
+	extraQuery, ok := getQueryFromExtra(extra)
+	if ok {
+		discoverQueries = append(discoverQueries, extraQuery)
+	}
+	// worth noting that the discovered hosts are not guaranteed to be rtsp camera.
+	// We assume that the user knew what they were looking for.
+	hosts, hostmap, err := viamupnp.FindHost(ctx, dis.logger, discoverQueries, dis.rootOnly)
+	if err != nil {
+		return nil, err
+	}
+
+	for hostNum, host := range hosts {
+		query := hostmap[host]
+		endpoints, ok := dis.endpointMap[query]
+		// the query had no endpoints
+		if !ok {
+			camConfig, err := createCameraConfig(createCameraName(hostNum, -1, query), "rtsp://"+host, query)
+			if err != nil {
+				return nil, err
+			}
+			cams = append(cams, camConfig)
+			continue
+		}
+		for endpointNum, endpoint := range endpoints {
+			camConfig, err := createCameraConfig(createCameraName(hostNum, endpointNum, query),
+				fmt.Sprintf("rtsp://%s:%s", host, endpoint), query)
+			if err != nil {
+				return nil, err
+			}
+			cams = append(cams, camConfig)
+		}
+	}
+
+	return cams, nil
+}
+
+// regex to remove non alpha numerics.
+var reg = regexp.MustCompile("[^a-zA-Z0-9]+")
+
+// createCameraName creates a camera name based on the query.
+func createCameraName(hostNum, endpointNum int, query viamupnp.DeviceQuery) string {
+	camName := fmt.Sprintf("camera%v", hostNum)
+	if endpointNum != -1 {
+		camName = fmt.Sprintf("%s-endpoint%v", camName, endpointNum)
+	}
+	if stripManufacturer := reg.ReplaceAllString(query.Manufacturer, ""); stripManufacturer != "" {
+		camName = fmt.Sprintf("%s-%s", camName, stripManufacturer)
+	}
+	if stripModel := reg.ReplaceAllString(query.ModelName, ""); stripModel != "" {
+		camName = fmt.Sprintf("%s-%s", camName, stripModel)
+	}
+	if stripSerial := reg.ReplaceAllString(query.SerialNumber, ""); stripSerial != "" {
+		camName = fmt.Sprintf("%s-%s", camName, stripSerial)
+	}
+	return camName
+}
```

Comment:
Is this the same as the onvif discover service's?

Reply:
nope, onvif doesn't explicitly call out which endpoint is which. 

the naming is also different since we are less likely to have all of the model/serial/manufacturer info. In the future we could try getting upnp discovery to match onvif, but that would require more changes in eliots repo.